python/pyspark/ml/torch/distributor.py

```diff
@@ -0,0 +1,491 @@
+            how CPU-specific code looks like.
+        spark : Optional[SparkSession], optional
+            An optional parameter that allows users to pass in a custom SparkSession argument
+            with a custom conf, by default None
+
+        Raises
+        ------
+        ValueError
+            If any of the parameters are incorrect.
+        """
+        super().__init__(num_processes, local_mode, use_gpu, spark)
+        self.framework = framework
+        self.ssl_conf = "pytorch.spark.distributor.ignoreSsl"  # type: ignore
+        self._validate_input_params()
+
+    def _validate_input_params(self) -> None:
+        """Validates input params
+
+        Raises
+        ------
+        ValueError
+            Thrown when user fails to provide correct input params
+        """
+        super()._validate_input_params()
+        if self.framework not in self.available_frameworks:
+            raise ValueError(
+                f"{self.framework} is not a valid framework."
+                f"Available frameworks: {self.available_frameworks}"
+            )
+
+    def _run_local_training(
+        self,
+        framework_wrapper_fn: Optional[Callable],
+        train_fn: Union[Callable, str],
+        *args: Any,
+    ) -> Optional[Any]:
+        CUDA_VISIBLE_DEVICES = "CUDA_VISIBLE_DEVICES"
+        old_cuda_visible_devices = os.environ.get(CUDA_VISIBLE_DEVICES, "")
+        cuda_state_was_set = CUDA_VISIBLE_DEVICES in os.environ
+        if not framework_wrapper_fn:
+            raise RuntimeError("Unknown combination of parameters")
+        try:
+            if self.use_gpu:
+                if "gpu" not in self.sc.resources:
+                    raise ValueError('"gpu" was not found in the context resources.')
+                gpus_owned = get_gpus_owned(self.sc.resources["gpu"].addresses)
+
+                if self.num_processes > len(gpus_owned):
+                    raise ValueError(
+                        f"""{self.num_processes} processes were requested
+                        for local training with GPU training but only
+                        {len(gpus_owned)} GPUs were available."""
+                    )
+                selected_gpus = [str(e) for e in random.sample(gpus_owned, self.num_processes)]
+                os.environ[CUDA_VISIBLE_DEVICES] = ",".join(selected_gpus)
+
+            output = framework_wrapper_fn(self.__dict__, train_fn, *args)
+        finally:
+            if cuda_state_was_set:
+                os.environ[CUDA_VISIBLE_DEVICES] = old_cuda_visible_devices
+            else:
+                if (os.getenv(CUDA_VISIBLE_DEVICES, "")) != "":
+                    del os.environ[CUDA_VISIBLE_DEVICES]
+
+        return output
+
+    @staticmethod
+    def _run_training_on_pytorch_file(
+        input_params: dict[str, Any], train_path: str, *args: Any
+    ) -> None:
+        training_command = create_torchrun_command(input_params, train_path, *args)
+        execute_command(training_command)
```

Comment:
The torchrun children process should output a model (to disk or return a pickled model data?) and then the `Distributor.run` should load the model and return it ?

Reply:
Yes, but to my understanding, that only happens when the input is a function. When the input is a file, the user code is expected to save the data themselves. There is no return value in these situations. Here's what I used for reference:
1. [PyTorch recommendation on how to create train.py files](https://pytorch.org/docs/stable/elastic/train_script.html) 
2. [General (External) guide](https://lambdalabs.com/blog/multi-node-pytorch-distributed-training-guide)

Reply:
Then I have a question,
if we running code on cluster mode spark cluster, then the "torchrun" subprocess will be executed in remote spark worker, then its output is also on remote worker disk, how can driver side get the output ?

Note that you cannot control which remote worker the spark task is scheduled to run on. This is very inconvenient to user. Shall we force user to output the data to centralized storage like dbfs ?

Reply:
We have 2 solutions for this:

1) Force user code output model to specified path, then Distributor code can pickle model data and collect it to driver by RDD.collect()

2) (databricks only) Force user code output model to specified path, then Distributor code copy model to a specified dbfs directory that is accessible from driver side.

Reply:
Yes, we most likely need to force user to output the data to centralized storage.

Reply:
> I assumed that if we tell them to store it within hdfs (or dbfs) then the driver could see that as well. 

The assumption is correct.
We should document this.

Reply:
+1

Reply:
Should this just be in the official documentation or could it be inside some docstring in this file?

Reply:
We should add it in official doc.
and, we should also provide a databricks doc page with guide that contains the usage of copying data to dbfs.